token-limiter-concurrent/test/Control/Concurrent/TokenLimiter/ConcurrentSpec.hs

```diff
@@ -189,6 +236,26 @@ spec = do
           nanos <- time_ $ replicateConcurrently_ numberOfThreads (waitDebit limiter 1)
           nanos `shouldSatisfy` (>= fromIntegral numberOfThreads * 10_000_000)
 
+      it "waits fairly" $ do
+        forAll (sized pure) $ \numberOfThreads -> do
+          queue <- newTQueueIO
+          let config =
+                TokenLimitConfig
+                  { tokenLimitConfigInitialTokens = 0,
+                    tokenLimitConfigMaxTokens = 1,
+                    tokenLimitConfigTokensPerSecond = 50
+                  }
+          limiter <- makeTokenLimiter config
+          let l :: [Int]
+              l = [1 .. numberOfThreads]
+          -- The threads start waiting in order, so they need to wake up in order
+          forConcurrently_ l $ \ix -> do
+            threadDelay (10_000 * ix)
```

Comment:
Didn't check these tests, random note: I found it can be useful to execute `stress` utility with CPU/memory/IO hogs while performing the concurrent tests repeatedly, scheduling errors are much proner to appear under high load.

Reply:
Great tip!

I tried this for a while and didn't see anything fail, but I probably didn't stress the system in exactly the right way to cause issues.